include/merlin/core_kernels.cuh

```diff
@@ -953,6 +982,175 @@ __forceinline__ __device__ void upsert_kernel_with_io_core(
+          bucket->keys[key_pos].load(cuda::std::memory_order_relaxed);
+
+      found_vote = g.ballot(insert_key == current_key);
+
+      if (found_vote) {
+        src_lane = __ffs(found_vote) - 1;
+        key_pos = (start_idx + tile_offset + src_lane) & (bucket_max_size - 1);
+        if (rank == src_lane) {
+          update_meta(bucket, key_pos, metas, key_idx);
+        }
+        if (local_size >= bucket_max_size) {
+          refresh_bucket_meta<K, V, M, TILE_SIZE>(g, bucket, bucket_max_size);
+        }
+        lock<Mutex, TILE_SIZE, true>(g, table->locks[bkt_idx]);
+        copy_vector<V, TILE_SIZE>(g, insert_value,
+                                  bucket->vectors + key_pos * dim, dim);
+        unlock<Mutex, TILE_SIZE, true>(g, table->locks[bkt_idx]);
+
+        break;
+      }
+      const unsigned empty_vote =
+          g.ballot(current_key == static_cast<K>(EMPTY_KEY) ||
+                   current_key == static_cast<K>(RECLAIM_KEY));
+      if (empty_vote) {
+        src_lane = __ffs(empty_vote) - 1;
+
+        key_pos = (start_idx + tile_offset + rank) & (bucket_max_size - 1);
+        AtomicKey<K>* current_atomic_key = &(bucket->keys[key_pos]);
+
+        if (rank == src_lane) {
+          occupy_result = try_occupy<K, V, M, TILE_SIZE>(g, bucket, insert_key,
+                                                         current_atomic_key);
+        }
+
+        occupy_result = g.shfl(occupy_result, src_lane);
+
+        if (occupy_result == OccupyResult::OCCUPIED_EMPTY ||
+            occupy_result == OccupyResult::OCCUPIED_RECLAIMED) {
+          key_pos =
+              (start_idx + tile_offset + src_lane) & (bucket_max_size - 1);
+          if (rank == src_lane) {
+            update_meta(bucket, key_pos, metas, key_idx);
+            local_size = atomicAdd(&(buckets_size[bkt_idx]), 1) + 1;
+          }
+          local_size = g.shfl(local_size, src_lane);
+
+          if (local_size >= bucket_max_size) {
+            refresh_bucket_meta<K, V, M, TILE_SIZE>(g, bucket, bucket_max_size);
+          }
+          lock<Mutex, TILE_SIZE, true>(g, table->locks[bkt_idx]);
+          copy_vector<V, TILE_SIZE>(g, insert_value,
+                                    bucket->vectors + key_pos * dim, dim);
+          unlock<Mutex, TILE_SIZE, true>(g, table->locks[bkt_idx]);
+          break;
+        } else if (occupy_result == OccupyResult::DUPLICATE) {
+          break;
+        } else if (occupy_result == OccupyResult::CONTINUE) {
+          continue;
+        }
+      }
+      tile_offset += TILE_SIZE;
+    }
+
+    while (occupy_result == OccupyResult::CONTINUE ||
+           occupy_result == OccupyResult::INITIAL) {
+      key_pos = bucket->min_pos.load(cuda::std::memory_order_acquire);
+      src_lane = (key_pos % TILE_SIZE);
+      OverrideResult override_result{OverrideResult::INITIAL};
+
+      if (rank == src_lane) {
+        const M min_meta =
+            bucket->min_meta.load(cuda::std::memory_order_acquire);
+        // TODO
```

Comment:
I think the TODO is unnecessary, for potential performance consideration, a separate `try_swap_min_meta` works for me.